config/genesis.go

```diff
@@ -0,0 +1,145 @@
+package config
+
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+
+	apiConfig "github.com/spacemeshos/go-spacemesh/api/config"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/hash"
+)
+
+const (
+	ExtraDataLen                 = 255
+	DefaultGenesisConfigFileDir  = "data"
+	DefaultGenesisConfigFileName = "genesis.conf"
+)
+
+type GenesisConfig struct {
+	Accounts    *apiConfig.GenesisAccountConfig `mapstructure:"genesis-api"`
+	GenesisTime string                          `mapstructure:"genesis-time"`
+	ExtraData   string                          `mapstructure:"genesis-extradata"`
+	GoldenATXID string
+	GenesisID   string
+	Path        string
+}
+
+func defaultGenesisConfig() *GenesisConfig {
+	return &GenesisConfig{
+		Accounts:    apiConfig.DefaultGenesisAccountConfig(),
+		GenesisTime: DefaultGenesisTime(),
+		GoldenATXID: DefaultGoldenATXId().Hash32().Hex(),
+		ExtraData:   DefaultGenesisExtraData(),
+		Path:        DefaultGenesisDataDir(),
+	}
+}
+
+func DefaultTestnetGenesisConfig() *GenesisConfig {
+	return &GenesisConfig{
+		Accounts: &apiConfig.GenesisAccountConfig{
+			Data: map[string]uint64{
+				"stest1qqqqqqygdpsq62p4qxfyng8h2mm4f4d94vt7huqqu9mz3": 100000000000000000,
+				"stest1qqqqqqylzg8ypces4llx4gnat0dyntqfvr0h6mcprcz66": 100000000000000000,
+				"stest1qqqqqq90akdpc97206485eu4m0rmacd3mxfv0wsdrea6k": 100000000000000000,
+				"stest1qqqqqq9jpsarr7tnyv0qr0edddwqpg3vcya4cccauypts": 100000000000000000,
+				"stest1qqqqqq8lpq7f5ghqt569nvpl8kldv8r66ms2yzgudsd5t": 100000000000000000,
+			},
+		},
+		GenesisTime: DefaultTestnetGenesisTime(),
+		ExtraData:   DefaultTestnetGenesisExtraData(),
+		GoldenATXID: DefaultTestnetGoldenATXId().Hash32().Hex(),
+		GenesisID:   CalcGenesisId([]byte(DefaultTestnetGenesisExtraData()), DefaultTestnetGenesisTime()).Hex(),
+		Path:        DefaultGenesisDataDir(),
+	}
+}
+
+func DefaultTestGenesisConfig() *GenesisConfig {
+	return &GenesisConfig{
+		Accounts:    apiConfig.DefaultTestGenesisAccountConfig(),
+		GenesisTime: DefaultTestGenesisTime(),
+		ExtraData:   DefaultTestGenesisExtraData(),
+		GoldenATXID: DefaultTestGoldenATXId().Hash32().Hex(),
+		GenesisID:   CalcGenesisId([]byte(DefaultTestGenesisExtraData()), DefaultTestGenesisTime()).Hex(),
+		Path:        DefaultGenesisDataDir(),
+	}
+}
+
+func (gc *GenesisConfig) Compare(payload *GenesisConfig) error {
+	if diff := cmp.Diff(payload, gc); diff != "" {
+		fmt.Printf("config files mismatch (-want +got):\n%s", diff)
```

Comment:
rm

Reply:
I added this on purpose to see the diff in a clean way when error occurs. If we include it into error printed then JSON log escapes it and it doesn't look good to understand. Or no point to print diff at all?